### PR TITLE
Fix bug on elastic search

### LINKED ISSUE #12791: [Bug] Code search failed with Elastic Search and Chinese characters
- Gitea version (or commit ref): 1.13.0+dev-580-gc6e4bc53a
- Git version: 2.11.0
- Operating system: alpine
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
  - [ ] Not relevant
- Log gist:

https://gist.github.com/jnlin/b55388dbc5e65890d87fa11f5701e523

## Description

Panic on source code line https://github.com/go-gitea/gitea/blob/master/modules/indexer/code/elastic_search.go#L274

## Screenshots
截圖 2020-09-10 下午2 20 55

### PR DESCRIPTION
**Pull request by lunny**
When input two words or some Unicode texts on the search box, the default elastic search highlighting plugin will return content with high light html tags. i.e. `<em>word1</em> <em>word2</em>`. 
But the previous codes will try to find `<em>word1 word2</em>` and return `startIndex = endIndex = -1` which will resutl in a panic.

Should fix #12791 

Since the new mapping option `"term_vector": "with_positions_offsets"` added, you have to delete your index previously.
You can try
```
curl -XDELETE http://localhost:9200/gitea_codes.v1
```
And restart gitea, the index will be rebuilt.

The new option will increse the space of the index but get faster on returning words positions.

### CODE REVIEW

File: modules/indexer/code/elastic_search.go [hunk 1]
```diff
@@ -251,6 +252,32 @@ func (b *ElasticSearchIndexer) Delete(repoID int64) error {
 	return err
 }
 
+func indexPos(content, start, end string) (int, int) {
```
Comment (zeripath) [inline modules/indexer/code/elastic_search.go:259]: Could you add a comment to explain what it is trying to do. It's really not clear what it's supposed to be doing so I can't really assess if it's doing it correctly.

**Comment by lunny** [comment]
@zeripath done.

File: modules/indexer/code/elastic_search.go [hunk 2]
```diff
@@ -251,6 +252,36 @@ func (b *ElasticSearchIndexer) Delete(repoID int64) error {
 	return err
 }
 
+// indexPos find words positions for start and the following end on content. It will
+// return the beginning position of the frist start and the ending position of the
```
Comment (lafriks) [inline modules/indexer/code/elastic_search.go:256]:
```suggestion
// return the beginning position of the first start and the ending position of the
```

**Review by lafriks — APPROVED** [review]

File: modules/indexer/code/elastic_search.go [hunk 3]
```diff
@@ -251,6 +252,36 @@ func (b *ElasticSearchIndexer) Delete(repoID int64) error {
...
+			if i+len(start) > len(content) {
+				return -1, -1
+			}
+			if content[i:i+len(start)] == start {
+				startIdx = i
+			}
+		}
+	}
+	return -1, -1
+}
```
Comment (zeripath) [inline modules/indexer/code/elastic_search.go]: So strings.Index(...) is actually quicker than iterating through the indices 

```suggestion
func indexPos(content, start, end string) (int, int) {
	if len(content) < len(start)+len(end) {
		return -1, -1
	}
	startIdx := strings.Index(content, start)
	if startIdx < 0 {
		return -1, -1
	}
	pos := startIdx + len(start)
	endIdx := strings.Index(content[pos:], end)
	if endIdx < 0 {
		return -1, -1
	}
	
	return startIdx, pos+endIdx
}
```

**Comment by lunny** [comment]
@zeripath done.

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
So need to figure out what happens with <em> content

**Comment by lunny** [comment]
@zeripath Updated on the first comment of mine.

**Pull request merged**